test/common/filter/config_discovery_impl_test.cc

```diff
@@ -127,311 +119,364 @@ config_source: { ads: {} }
+          Network::ListenerFilterFactoryCb, Server::Configuration::ListenerFactoryContext,
+          TcpListenerFilterConfigProviderManagerImpl, Envoy::ProtobufWkt::Struct> {
+public:
+  const std::string getTypeUrl() const override { return "google.protobuf.Struct"; }
+  const std::string getFilterType() const override { return "listener"; }
+  const std::string getConfigReloadCounter() const override {
+    return "extension_config_discovery.tcp_listener_filter.foo.config_reload";
+  }
+  const std::string getConfigFailCounter() const override {
+    return "extension_config_discovery.tcp_listener_filter.foo.config_fail";
+  }
+};
+
+// UDP listener filter test
+class UdpListenerFilterConfigDiscoveryImplTest
+    : public FilterConfigDiscoveryImplTest<
+          Network::UdpListenerFilterFactoryCb, Server::Configuration::ListenerFactoryContext,
+          UdpListenerFilterConfigProviderManagerImpl,
+          test::integration::filters::TestUdpListenerFilterConfig> {
+public:
+  const std::string getTypeUrl() const override {
+    return "test.integration.filters.TestUdpListenerFilterConfig";
+  }
+  const std::string getFilterType() const override { return "listener"; }
+  const std::string getConfigReloadCounter() const override {
+    return "extension_config_discovery.udp_listener_filter.foo.config_reload";
+  }
+  const std::string getConfigFailCounter() const override {
+    return "extension_config_discovery.udp_listener_filter.foo.config_fail";
+  }
+};
+
+/***************************************************************************************
+ *                  Parameterized test for                                             *
+ *     HTTP filter,  TCP listener filter And UDP listener filter                       *
+ *                                                                                     *
+ ***************************************************************************************/
+template <typename FilterConfigDiscoveryTestType>
+class FilterConfigDiscoveryImplTestParameter : public testing::Test {
+public:
+  FilterConfigDiscoveryTestType config_discovery_test_;
+};
+
+// The test filter types.
+using FilterConfigDiscoveryTestTypes =
+    ::testing::Types<HttpFilterConfigDiscoveryImplTest, TcpListenerFilterConfigDiscoveryImplTest,
+                     UdpListenerFilterConfigDiscoveryImplTest>;
+
+TYPED_TEST_SUITE(FilterConfigDiscoveryImplTestParameter, FilterConfigDiscoveryTestTypes);
+
+// TYPED_TEST will run the same test for each of the above filter type.
+TYPED_TEST(FilterConfigDiscoveryImplTestParameter, DestroyReady) {
+  this->config_discovery_test_.setup();
+  EXPECT_CALL(this->config_discovery_test_.init_watcher_, ready());
 }
 
-TEST_F(FilterConfigDiscoveryImplTest, Basic) {
+TYPED_TEST(FilterConfigDiscoveryImplTestParameter, Basic) {
   InSequence s;
-  setup();
-  EXPECT_EQ("foo", provider_->name());
-  EXPECT_EQ(absl::nullopt, provider_->config());
+  this->config_discovery_test_.setup();
```

Comment:
nit: no need for `this->`

Reply:
Without "this->", it won't compile. Please check below:

https://google.github.io/googletest/advanced.html#:~:text=//%20us%20to%20visit%20the%20members%20of%20FooTest%20via%20%27this%27.%0A%20%20TypeParam%20n%20%3D%20this%2D%3Evalue_%3B